sound/soc/sof/intel/tgl.c

```diff
@@ -50,6 +50,24 @@ static int tgl_dsp_core_put(struct snd_sof_dev *sdev, int core)
 	return 0;
 }
 
+static int tgl_ipc4_dsp_core_put(struct snd_sof_dev *sdev, int core)
+{
+	const struct sof_ipc_pm_ops *pm_ops = sdev->ipc->ops->pm;
+	int ret;
+
+	if (pm_ops->set_core_state) {
+		ret = pm_ops->set_core_state(sdev, core, false);
+		if (ret < 0)
+			return ret;
+	}
+
+	/* power down primary core and return */
+	if (core == SOF_DSP_PRIMARY_CORE)
+		return hda_dsp_core_reset_power_down(sdev, BIT(core));
+
+	return 0;
+}
+
```

Comment:
I don't fully understand how the .get is common with IPC3 but the .put is not - it's specific to IPC4.

That makes me a bit nervous. Is this difference intentional, and if yes can we have a comment in the commit message and/or the code to explain how this works?

Reply:
sure, I will update commit message. IPC3  doesn't save context into imr when powering off core 0 but ipc4 does. FW doesn't need to do anything for core 0 for tgl or later platform.